pillar/zedrouter: introduce legacy MAC address generation mode

rouming opened the pull request:
The legacy MAC generation is not "that" random and probability of repeating MAC addresses across nodes is high. Later the generation algorithm was changed and more randomness was introduced, but some applications may be already configured with already allocated MAC address, and MAC address change on EVE node upgrade (switch from old generation logic to new one) can cause problems with the guest network.

This commit introduces old (but still so beloved) legacy MAC address generation, by setting the 'zedrouter.legacy.mac.address' to true.

This can help eliminate guest network misconfiguration on EVE node upgrades for certain cases by setting the 'zedrouter.legacy.mac.address' to true in advance, prior the upgrade.

The legacy algorithm was extracted from code prior the following patch: fdd167efa627 ("Generate random MAC addresses even for local network instances")

rene commented on pkg/pillar/types/global.go:
```diff
@@ -284,6 +284,14 @@ const (
+	// already configured with already allocated MAC address, and MAC address
+	// change on EVE node upgrade (switch from old generation logic to new one)
+	// can cause problems with the guest network.
+	LegacyMACAddress GlobalSettingKey = "zedrouter.legacy.mac.address"
```
@rouming , please, add some documentation about this new option here: https://github.com/lf-edge/eve/blob/master/docs/CONFIG-PROPERTIES.md

eriknordmark replied:
Also as part of the comment and docs please mention that this is for local network instances.

milan-zededa commented on pkg/pillar/cmd/zedrouter/ipam.go:
```diff
@@ -68,6 +91,17 @@ func (z *zedrouter) generateAppMac(appUUID uuid.UUID, adapterNum int, appNum int
 
+func (z *zedrouter) generateAppMac(appUUID uuid.UUID, adapterNum int, appNum int,
+	netInstStatus *types.NetworkInstanceStatus) net.HardwareAddr {
+	if z.legacyMACAddr {
```
One important thing to note is that we have not changed how MAC addresses are generated for switch network instances, only for local (see the code before and after, lines are rearranged but the logic have not changed for the switch case). Meaning that this `if z.legacyMACAddr` condition could be under `case types.NetworkInstanceTypeLocal` and I would also avoid this log messages about legacy/random MAC address for switch NI.
Moreover, I would suggest to name the config option: `network.local.legacy.mac.address` (similar to `network.switch.enable.arpsnoop` which we have for switch NI). Also, I would not put `zedrouter` there, as we do not have names of microservices in other config properties either.

rouming replied:
Thanks, will change accordingly.

milan-zededa reviewed (commented):
Overall LGTM, only some suggestions added. Thanks @rouming for taking care of this!
Later we should also improve this by applying change in this new config property only to new or purged applications, not to existing. We could perhaps persist MAC address, or just its version, alongside `appNum` and other things that are persisted for app using pubsub. This could help to prevent similar problems in the future.

milan-zededa commented:
Note that adding a new config property requires update of `TestNewConfigItemSpecMap` unit test.

rouming commented:
Closed accidentally, wrong button.

rouming commented:
Difference to the previous version:
1. Addressed @milan-zededa comments: code tweaks, setting rename, unit test tweaks.
2. Addressed @rene comments: a proper line in the doc.
3. Addressed @eriknordmark comments: comment and doc mention "local network instances".

eriknordmark reviewed (approved):
LGTM
